crates/executor/src/syscalls/eip4788.rs

```diff
@@ -5,60 +5,23 @@ use crate::{
     errors::{ExecutorError, ExecutorResult},
     TrieDBProvider,
 };
-use alloc::{boxed::Box, vec::Vec};
+use alloc::boxed::Box;
 use alloy_eips::eip4788::BEACON_ROOTS_ADDRESS;
-use alloy_primitives::{Address, Bytes, B256, U256};
+use alloy_primitives::B256;
 use kona_mpt::TrieHinter;
 use op_alloy_genesis::RollupConfig;
-use op_alloy_rpc_types_engine::OpPayloadAttributes;
-use revm::{
-    db::State,
-    primitives::{
-        BlockEnv, CfgEnvWithHandlerCfg, Env, EnvWithHandlerCfg, OptimismFields, TransactTo, TxEnv,
-    },
-    DatabaseCommit, Evm,
-};
-
-/// Execute the EIP-4788 pre-block beacon root contract call.
-pub(crate) fn pre_block_beacon_root_contract_call<F, H>(
-    db: &mut State<&mut TrieDB<F, H>>,
-    config: &RollupConfig,
-    block_number: u64,
-    initialized_cfg: &CfgEnvWithHandlerCfg,
-    initialized_block_env: &BlockEnv,
-    payload: &OpPayloadAttributes,
-) -> ExecutorResult<()>
-where
-    F: TrieDBProvider,
-    H: TrieHinter,
-{
-    // apply pre-block EIP-4788 contract call
-    let mut evm_pre_block = Evm::builder()
-        .with_db(db)
-        .with_env_with_handler_cfg(EnvWithHandlerCfg::new_with_cfg_env(
-            initialized_cfg.clone(),
-            initialized_block_env.clone(),
-            Default::default(),
-        ))
-        .build();
-
-    // initialize a block from the env, because the pre block call needs the block itself
-    apply_beacon_root_contract_call(
-        config,
-        payload.payload_attributes.timestamp,
-        block_number,
-        payload.payload_attributes.parent_beacon_block_root,
-        &mut evm_pre_block,
-    )
-}
+use reth_evm::ConfigureEvm;
+use revm::{db::State, DatabaseCommit, Evm};
 
 /// Apply the EIP-4788 pre-block beacon root contract call to a given EVM instance.
-fn apply_beacon_root_contract_call<F, H>(
+pub(crate) fn apply_beacon_root_contract_call<F, H, C: ConfigureEvm>(
     config: &RollupConfig,
+    evm_config: &C,
     timestamp: u64,
     block_number: u64,
     parent_beacon_block_root: Option<B256>,
-    evm: &mut Evm<'_, (), &mut State<&mut TrieDB<F, H>>>,
+    // ZTODO: understand default external context better, is it ok to allow any here?
+    evm: &mut Evm<'_, C::DefaultExternalContext<'_>, &mut State<&mut TrieDB<F, H>>>,
```

Comment:
@clabby I think this change is fine to allow the Evm to be parameterized by any DefaultExternalContext rather than forcing it to be `()`. Do you see any problem with that?

Reply:
This is all good. Right now the enforcement on `()` limits a degree of freedom that consumers might want.